Refactor PICMI custom user input (#4996)
ComputationalRadiationPhysics/picongpu: docs/source/usage/picmi/intro.rst

.. _PICMI:
Intro
=====

PICMI is a python interface for configuring and running PIC-simulations, defined via the python reference implementation
available from `PICMI-github <https://github.com/picmi-standard/picmi>` and via `pypip <https://pypi.org/project/picmistandard/>`.

PICMI allows a user to configure a simulation in a python script, the user script, by building a simulation python object from different
standardised building blocks.

.. figure::  media/PICMI_structure_DetailMedium.png
   :name: PICMI_mediumDetail

   general overview of PICMI interface, see python implementation for details

> From this simulation object a user may then generate input files for all PIC-simulation codes supporting PICMI,but some

Reviewer: ```suggestion
From this simulation object a user may then generate input files for all PIC-simulation codes supporting PICMI, but some
```

Author: fixed

features may be code specific.

Usage Quick guide
-----------------

> To use the PICMI interface you need working a picongpu environment, see :ref:`_install-pyth` and :ref:`the Setup part of PIConGPU in 5 Minutes on Hemera <hemeraIn5min>` for instructions.

Reviewer: ```suggestion
To use the PICMI interface you need a working picongpu environment, see :ref:`_install-pyth` and :ref:`the Setup part of PIConGPU in 5 Minutes on Hemera <hemeraIn5min>` for instructions.
```

Author: fixed

In addition you need to install the python dependencies of the PIConGPU PICMI implementation to your python environment.

.. code:: shell

  pip install -r $PIC_SRC/lib/python/piconpgu/picmi/requirements.txt

.. note::

  If you are using a python package manager other than pip, you may instead install all the requirements listed in ``$PIC_SRC/lib/python/piconpgu/picmi/requirements.txt`` and ``$PIC_SRC/lib/python/piconpgu/pypicongpu/requirements.txt``

And include the PIConGPU PICMI implementation in your ``PYHTONPATH`` environment variable, for example by

.. code:: shell

  export PYTHONPATH=$PICSRC/lib/python:$PYTHONPATH

.. note::

  If you are using one of our pre-configured profiles, this is automaticlly done when you source a profile

.. note::
   Above, we used:

>    -  ``$PICSRC`` as a short hand for the path to picongpu's source code directory, usually provided from your shell environment

Reviewer: ```suggestion
   -  ``$PICSRC`` as a short hand for the path to picongpu's source code directory, provided from your shell environment if a per-configured profile is used.
```

Reviewer 2: ```suggestion
   -  ``$PICSRC`` as a short hand for the path to picongpu's source code directory, usually provided from your shell environment if a pre-configured profile is used.
```
Correcting a typo in the above.

Author: good suggestion

PIConGPU specifics
------------------

PIConGPU carries with it an implementation of the PICMI standard, with picongpu specific extensions.

> Therefore names of classes differ a little from the standard namings and sometimes class provide additional options.

Reviewer: ```suggestion
Therefore names of classes differ a little from the standard namings and sometimes a class may provide additional options.
```

Author: fixed

Usage Quick-start
-----------------

To use the PICMI interface you need a working picongpu environment, see :ref:`install instructions <install-path>` and :ref:`the Setup part of PIConGPU in 5 Minutes on Hemera <hemeraIn5min>` for instructions.
In addition you need to install the python dependencies of the PIConGPU PICMI implementation to your python environment.

To install the python dependencies you may either run the command below or install all the requirements listed in ``$PIC_SRC/lib/python/piconpgu/picmi/requirements.txt`` and ``$PIC_SRC/lib/python/piconpgu/pypicongpu/requirements.txt``.

.. code:: shell

  pip install -r $PIC_SRC/lib/python/piconpgu/picmi/requirements.txt

.. note::

After you have installed the dependencies you must include the PIConGPU PICMI implementation in your ``PYHTONPATH`` environment variable, for example by

.. code:: shell

  export PYTHONPATH=$PICSRC/lib/python:$PYTHONPATH

.. note::

  If you are using one of our pre-configured profiles, this is done automatically when you source a profile

.. note::
   Above, we used:

>    -  ``$PICSRC`` as a short hand for the path to picongpu's source code directory, provided from your shell environment if a per-configured profile is used.

Reviewer 2: In rendering there is a very big gap. You might want to close it (haven't thought about how).

Author: not really possible, this is not Latex unfortunately

.. _PICMI:
Intro
=====

PICMI is a python interface for configuring and running PIC-simulations, defined via the python reference implementation

> available `PICMI-github <https://github.com/picmi-standard/picmi>`_ and via `pypip <https://pypi.org/project/picmistandard/>`_.

Reviewer 2: Link doesn't render correctly.

Author: fixed

PICMI allows a user to configure a simulation in a python script, called user script, by building a simulation python object from different
standardised building blocks.

.. image::  ../media/PICMI_structure_DetailMiddle.png

>    :name: PICMI_mediumDetail

Reviewer 2: Doesn't render correctly.

Author: fixed

Author: for future reference, setting the width or similar is mandatory for figures, default is 0 ...

   general overview of PICMI interface, see python implementation for details

From this simulation object a user may then generate input files for all PIC-simulation codes supporting PICMI, but some
features may be code specific.

Usage Quick-start
-----------------

> To use the PICMI interface you need a working picongpu environment, see :ref:`_install-pyth` and :ref:`the Setup part of PIConGPU in 5 Minutes on Hemera <hemeraIn5min>` for instructions.

Reviewer 2: `install_pyth` doesn't render correctly.

Author: typo, fixed

> To use the PICMI interface you need a working picongpu environment, see :ref:`_install-pyth` and :ref:`the Setup part of PIConGPU in 5 Minutes on Hemera <hemeraIn5min>` for instructions.

Reviewer 2: ```suggestion
To use the PICMI interface you need a working PIConGPU environment, see :ref:`_install-pyth` and :ref:`the Setup part of PIConGPU in 5 Minutes on Hemera <hemeraIn5min>` for instructions.
```
or
```suggestion
To use the PICMI interface you need a working `picongpu` environment, see :ref:`_install-pyth` and :ref:`the Setup part of PIConGPU in 5 Minutes on Hemera <hemeraIn5min>` for instructions.
```

Author: fixed

> In addition you need to install the python dependencies of the PIConGPU PICMI implementation to your python environment.

Reviewer 2: ```suggestion
In addition you need to install the Python dependencies of the PIConGPU PICMI implementation to your Python environment.
```
Everywhere...

Author: fixed

.. warning::

   We strive to quickly contribute these parameters to PICMI upstream,
   so this list is to be considered volatile.

- Simulation

  - ``__init__(..., picongpu_template_dir)``:
    Specify the template dir to use for code generation,
    please refer to :ref:`the documentation on the matter for details <picmi-custom-generation>`
  - ``__init__(...,  picongpu_custom_user_input)``:
    pass custom user input to the code generation. this may be used in conjunction with custom templates to change the code generation.
    See :ref:`PICMI custom code generation<picmi-custom-generation>` for the documentation on using custom input.
  - ``__init__(...,  picongpu_typical_ppc)`` typical ppc to be used for normalization in PIConGPU
  - ``write_input_file(..., pypicongpu_simulation)``:
    use a :ref:`PyPIConGPU simulation<PyPIConGPU_Intro>` object instead of an PICMI- simulation object to generate a PIConGPU input.
  - ``get_as_pypicongpu()``:
    convert the PICMI simulation object to an equivalent :ref:`PyPIConGPU <PyPIConGPU_Intro>` simulation object.
  - ``picongpu_get_runner()``:
    Retrieve a :ref:`PyPIConGPU Runner <pypicongpu-running>`

>  - Grid

Reviewer 2: What's going on with this?

Author: fixed rendering issue

  - ``picongpu_template_dir``:
    Specify the template dir to use for code generation,
    please refer to :ref:`the documentation on the matter for details <picmi-custom-generation>`
  - ``picongpu_custom_user_input``:
    pass custom user input to the code generation. this may be used in conjuntion with custom templates to change the code
    generation.

>     please refer to `missing Reference` for the docuemntation on using custom input

Reviewer: Missing a reference

Reviewer: It would be great if the warm plasma example (or some other piece of code) shows how the picongpu extensions are used, but it is not necessary for this PR

Author: missing reference fixed

Author: One of the next PR will replace the example with the LWFA example, using most picongpu extensions
